{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "# change all these to relative paths -- relative to the jupyter notebook, will need pathlib\n",
    "sys.path.extend([\n",
    "    \"C:\\\\Users\\\\Wyss User\\\\AppData\\\\Local\\\\Packages\\\\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\\\\LocalCache\\\\local-packages\\\\Python311\\\\site-packages\",\n",
    "    \"C:\\\\Users\\\\Wyss User\\\\Documents\\\\EVs\\\\OLINK\\\\src\",\n",
    "])\n",
    "\n",
    "from config import hgnc_ids, high_fractions, low_fractions, MISSING_FASTA_SEQUENCES, CELL_TYPES\n",
    "from brainrnaseq_specificity import map_hgnc_ids, create_enrichment_dataframe, process_hgnc_data, filter_low_tau\n",
    "from deeptmhmm_localization import identify_localization, parse_gz_file, get_localization_data\n",
    "from identify_targets import identify_targets, generate_protein_dataframe\n",
    "from olink_fractionation import analyze_fractionation\n",
    "from raw_data_preprocessing import clean_up_raw_data, plot_protein_fractionation, ev_association_score_df, find_ratio, calculate_fractionation_scores\n",
    "from specificity_functions import calculate_enrichment\n",
    "from gtex_specificity import gtex_specificity\n",
    "\n",
    "from io import StringIO\n",
    "\n",
    "import pandas as pd\n",
    "import requests\n",
    "\n",
    "os.chdir(\"C:\\\\Users\\\\Wyss User\\\\Documents\\\\EVs\\\\OLINK\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "assay_list_path = \"data\\\\231220_ht_panel_assay_list.xlsx\"\n",
    "brain_rna_seq_raw_path = \"data\\\\240411_brain_rna_seq_raw.csv\"\n",
    "output_directory = \"outputs\\\\ht_output\"\n",
    "plate_layout_path = \"data\\\\231204_Walt_Olink_HT_Plate.xlsx\"\n",
    "raw_data = \"data\\\\240214_Walt_Olink_HT_Raw.parquet\"\n",
    "uniprot_fasta_database = \"data\\\\uniprot_fasta_database.gz\"\n",
    "\n",
    "gtex_path = r\"C:\\\\Users\\\\Wyss User\\\\Documents\\\\EVs\\\\OLINK\\\\data\\\\GTEx_Analysis_v10_RNASeQCv2.4.2_gene_median_tpm.gct.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a tidy dataframe from the raw data file.\n",
    "tidy_data = clean_up_raw_data(raw_data, plate_layout_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SI Table 4\n",
    "\n",
    "# Load data\n",
    "brain_rna_seq = pd.read_csv(brain_rna_seq_raw_path)\n",
    "hgnc_uniprot_mapping_data = process_hgnc_data(hgnc_ids)\n",
    "brain_rna_seq = brain_rna_seq.merge(hgnc_uniprot_mapping_data, left_on=\"id\", right_on=\"hgnc_id\", how=\"inner\").dropna(subset=[\"uniprot_ids\"]).drop_duplicates(subset=[\"uniprot_ids\"])\n",
    "brain_rna_seq.set_index([\"uniprot_ids\", \"symbol\", \"name\", \"alias_symbol\", \"alias_name\"], inplace=True)\n",
    "expression_df = create_enrichment_dataframe(brain_rna_seq)\n",
    "\n",
    "# Identify low tau score proteins\n",
    "low_tau_score = filter_low_tau(expression_df)\n",
    "\n",
    "# Identify correct fractionation pattern proteins\n",
    "fractionation_uniprot_ids = analyze_fractionation(tidy_data, high_fractions, low_fractions, sample_health=\"healthy\", mean_median_individual=\"individual_median\")\n",
    "\n",
    "# Identify protein localizations\n",
    "fasta_sequences = parse_gz_file(uniprot_fasta_database)\n",
    "fasta_sequences.update(MISSING_FASTA_SEQUENCES)\n",
    "assays = pd.read_excel(assay_list_path)\n",
    "localization_ids = get_localization_data(assays, fasta_sequences, [\"TMhelix\", \"internal\", \"external\"], output_directory)\n",
    "\n",
    "# Generate categorized protein dataframes\n",
    "int_low_tau = generate_protein_dataframe(low_tau_score, fractionation_uniprot_ids, localization_ids[\"internal\"], \"internal\", tidy_data, high_fractions, low_fractions)\n",
    "tm_low_tau = generate_protein_dataframe(low_tau_score, fractionation_uniprot_ids, localization_ids[\"TMhelix\"], \"transmembrane\", tidy_data, high_fractions, low_fractions)\n",
    "ext_low_tau = generate_protein_dataframe(low_tau_score, fractionation_uniprot_ids, localization_ids[\"external\"], \"external\", tidy_data, high_fractions, low_fractions)\n",
    "\n",
    "# Concatenate results\n",
    "low_tau_with_fract_pattern = pd.concat([int_low_tau, tm_low_tau, ext_low_tau])\n",
    "\n",
    "# Re-name columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Wyss User\\Documents\\EVs\\OLINK\\src\\specificity_functions.py:31: RuntimeWarning: invalid value encountered in divide\n",
      "  row_x = array / max(array)\n"
     ]
    }
   ],
   "source": [
    "# SI Table 5 \n",
    "\n",
    "# Load and Process Brain RNA-Seq Data\n",
    "brain_rna_seq = map_hgnc_ids(hgnc_ids, brain_rna_seq_raw_path)\n",
    "expression_df = create_enrichment_dataframe(brain_rna_seq)\n",
    "\n",
    "# Define localization categories and identify proteins with the correct localization\n",
    "\n",
    "localization_types = [\"TMhelix\", \"internal\", \"external\"]\n",
    "localization_ids = {loc: identify_localization(assays, loc, output_directory) for loc in localization_types}\n",
    "\n",
    "# Store results for each localization category\n",
    "localization_results = []\n",
    "\n",
    "# Process each localization category\n",
    "for loc, uniprot_ids in localization_ids.items():\n",
    "    proteins_with_fract = set(uniprot_ids) & set(fractionation_uniprot_ids)\n",
    "\n",
    "    # Compute EV association scores in a vectorized manner\n",
    "    fractionation_scores = [\n",
    "        find_ratio(tidy_data[protein], high_fractions, low_fractions)\n",
    "        for protein in proteins_with_fract\n",
    "    ]\n",
    "\n",
    "    # Create a DataFrame\n",
    "    df = pd.DataFrame({\n",
    "        \"uniprot_ids\": list(proteins_with_fract),\n",
    "        \"ev_association_score\": fractionation_scores,\n",
    "    })\n",
    "\n",
    "    # Merge with gene mapping data\n",
    "    df = df.merge(hgnc_uniprot_mapping_data, on=\"uniprot_ids\")\n",
    "    df[\"localization\"] = loc\n",
    "\n",
    "    # Append to results\n",
    "    localization_results.append(df)\n",
    "\n",
    "# Concatenate results\n",
    "fract_pattern = pd.concat(localization_results, ignore_index=True)\n",
    "\n",
    "# Compute Tau Scores and Filter by Cutoff\n",
    "tau_score_cutoff = 0.75\n",
    "enrichment_values = expression_df.apply(lambda row: calculate_enrichment(row, \"tau\"), axis=1)\n",
    "high_tau_score = enrichment_values[enrichment_values > tau_score_cutoff]\n",
    "\n",
    "filtered_expression_df = expression_df[expression_df.index.isin(high_tau_score.index.tolist())]\n",
    "\n",
    "max_col_list = []\n",
    "uniprots_list = []\n",
    "\n",
    "for index, row in filtered_expression_df.iterrows():\n",
    "    max_column = row.idxmax()\n",
    "    max_col_list.append(max_column)\n",
    "    uniprots_list.append(index)\n",
    "\n",
    "cell_type_targets = pd.DataFrame({\"cell_type\": max_col_list, \"uniprot_id\": uniprots_list})\n",
    "\n",
    "si_table_5 = fract_pattern.merge(cell_type_targets, left_on = \"uniprot_ids\", right_on = \"uniprot_id\", how = \"inner\")\n",
    "high_tau_score = high_tau_score.reset_index()\n",
    "si_table_5 = si_table_5.merge(high_tau_score, on = \"uniprot_ids\")\n",
    "\n",
    "#GTex Specificity\n",
    "gtex_data = gtex_specificity(gtex_path)\n",
    "si_table_5 = si_table_5.merge(gtex_data, on = \"ensembl_gene_id\")\n",
    "\n",
    "# need to rename SI table columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate the data for figure 1d and 1e\n",
    "\n",
    "associated_proteins = []\n",
    "associated_samples = []\n",
    "associated_median_npx = []\n",
    "\n",
    "fig_1d_associated = [\"P08758\", \"P07355\", \"P09525\", \"Q9NP79\"]\n",
    "fig_1d_contaminant = [\"P02751\", \"P00734\", \"P36955\", \"P01024\"]\n",
    "\n",
    "tidy_data_sec = tidy_data[\n",
    "    (tidy_data.index.get_level_values(\"Health\") == \"Healthy\")\n",
    "    & (tidy_data.index.get_level_values(\"Sample\").str.contains(\"SEC\"))\n",
    "]\n",
    "\n",
    "for column in fig_1d_associated:\n",
    "    df = tidy_data_sec[column]\n",
    "\n",
    "    for sample in list(df.index.get_level_values(\"Sample\").unique()):\n",
    "        associated_proteins.append(column)\n",
    "        associated_samples.append(sample)\n",
    "        sample_df = df[df.index.get_level_values(\"Sample\") == sample]\n",
    "        associated_median_npx.append(sample_df.median())\n",
    "\n",
    "associated_proteins_df = pd.DataFrame({\"protein\": associated_proteins, \"sample\": associated_samples, \"median_npx\": associated_median_npx})\n",
    "\n",
    "contaminant_proteins = []\n",
    "contaminant_samples = []\n",
    "contaminant_median_npx = []\n",
    "\n",
    "for column in fig_1d_contaminant:\n",
    "    df = tidy_data_sec[column]\n",
    "\n",
    "    for sample in list(df.index.get_level_values(\"Sample\").unique()):\n",
    "        contaminant_proteins.append(column)\n",
    "        contaminant_samples.append(sample)\n",
    "        sample_df = df[df.index.get_level_values(\"Sample\") == sample]\n",
    "        contaminant_median_npx.append(sample_df.median())\n",
    "\n",
    "contaminant_proteins_df = pd.DataFrame({\"protein\": contaminant_proteins, \"sample\": contaminant_samples, \"median_npx\": contaminant_median_npx})\n",
    "\n",
    "associated_proteins = associated_proteins_df.pivot(columns = \"protein\", index = \"sample\")\n",
    "associated_proteins.to_excel(\"associated_proteins.xlsx\")\n",
    "\n",
    "contaminant_proteins = contaminant_proteins_df.pivot(columns = \"protein\", index = \"sample\")\n",
    "contaminant_proteins.to_excel(\"contaminant_proteins.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# identify targets found inside microglia cells that may be EV associated using the raw data file\n",
    "\n",
    "identify_targets(\n",
    "    assay_list_path = assay_list_path,\n",
    "    uniprot_fasta_database = uniprot_fasta_database, \n",
    "    brain_rna_seq_raw_path = brain_rna_seq_raw_path,\n",
    "    region = \"TMhelix\", \n",
    "    cell_type = \"oligodendrocyte\", \n",
    "    specificity_metric = \"tau\", \n",
    "    specificity_cutoff = 0.75,\n",
    "    high_fractions = ['9', '10'], \n",
    "    low_fractions = ['7', '11', '12', '13'], \n",
    "    sample_health = 'healthy', \n",
    "    mean_median_individual = 'individual_median',\n",
    "    raw_olink_data_file = raw_data,\n",
    "    plate_layout_dataframe = plate_layout_path,\n",
    "    output_directory = output_directory\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# identify targets found inside microglia cells that may be EV associated using the tidy dataframe\n",
    "\n",
    "identify_targets(\n",
    "    assay_list_path = assay_list_path,\n",
    "    uniprot_fasta_database = uniprot_fasta_database, \n",
    "    brain_rna_seq_raw_path = brain_rna_seq_raw_path,\n",
    "    region = \"internal\", \n",
    "    cell_type = \"microglia\", \n",
    "    specificity_metric = \"tau\", \n",
    "    specificity_cutoff = 0.75,\n",
    "    high_fractions = ['9', '10'], \n",
    "    low_fractions = ['7', '11', '12', '13'], \n",
    "    sample_health = 'healthy', \n",
    "    mean_median_individual = 'individual_median',\n",
    "    plate_layout_dataframe = plate_layout_path,\n",
    "    tidy_dataframe = tidy_data,\n",
    "    output_directory = output_directory\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a box-and-whisker plot for a target of interest\n",
    "\n",
    "plot_protein_fractionation(tidy_data, \"Q9Y251\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
